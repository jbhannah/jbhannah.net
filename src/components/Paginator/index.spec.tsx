import { render, screen } from "@testing-library/react"
import * as React from "react"
import { Paginator } from "."

const lis = (tree) => tree.querySelectorAll("li")
const linkAt = (tree, i) => lis(tree)[i].querySelector("a")
const hrefAt = (tree, i) => linkAt(tree, i).getAttribute("href")

const TestPaginator = (props) =>
  render(<Paginator data-testid="paginator" {...props} />)

describe("Paginator", () => {
  describe("with one page", () => {
    it("renders nothing", () => {
      const { container } = TestPaginator({
        base: "test",
        page: 1,
        numPages: 1,
      })

      expect(container).toMatchSnapshot()
    })
  })

  describe("with three pages", () => {
    describe("on the first page", () => {
      beforeEach(() => TestPaginator({ base: "/", page: 1, numPages: 3 }))

      it("renders correctly", () => {
        expect(screen.getByTestId("paginator")).toMatchSnapshot()
      })

      it("contains seven <li> elements", () => {
        expect(lis(screen.getByTestId("paginator"))).toHaveLength(7)
      })

      it("does not link to the first page", () => {
        expect(
          screen.getByTestId("paginator").querySelectorAll("a")
        ).toHaveLength(4)
        expect(linkAt(screen.getByTestId("paginator"), 0)).toBeNull()
        expect(linkAt(screen.getByTestId("paginator"), 1)).toBeNull()
        expect(linkAt(screen.getByTestId("paginator"), 2)).toBeNull()
      })

      it("links to the next page", () => {
        expect(hrefAt(screen.getByTestId("paginator"), 5)).toBe("/page/2")
      })

      it("links to the last page", () => {
        expect(hrefAt(screen.getByTestId("paginator"), 6)).toBe("/page/3")
      })
    })

    describe("on the last page", () => {
      beforeEach(() => TestPaginator({ base: "/", page: 3, numPages: 3 }))

      it("renders correctly", () => {
        expect(screen.getByTestId("paginator")).toMatchSnapshot()
      })

      it("links to the first page", () => {
        expect(hrefAt(screen.getByTestId("paginator"), 0)).toBe("/")
      })

      it("links to the previous page", () => {
        expect(hrefAt(screen.getByTestId("paginator"), 1)).toBe("/page/2")
      })

      it("does not link to the last page", () => {
        expect(linkAt(screen.getByTestId("paginator"), 4)).toBeNull()
        expect(linkAt(screen.getByTestId("paginator"), 5)).toBeNull()
        expect(linkAt(screen.getByTestId("paginator"), 6)).toBeNull()
      })
    })
  })
})
