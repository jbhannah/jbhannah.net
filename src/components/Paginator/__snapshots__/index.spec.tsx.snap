// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator with one page renders nothing 1`] = `<div />`;

exports[`Paginator with three pages on the first page renders correctly 1`] = `
<footer
  data-testid="paginator"
>
  <ul
    class="css-fm3157"
  >
    <li>
      «
    </li>
    <li>
      ‹
    </li>
    <li>
      1
    </li>
    <li>
      <a
        class="css-nk3r0j"
        href="/page/2"
      >
        2
      </a>
    </li>
    <li>
      <a
        class="css-nk3r0j"
        href="/page/3"
      >
        3
      </a>
    </li>
    <li>
      <a
        class="css-nk3r0j"
        href="/page/2"
      >
        ›
      </a>
    </li>
    <li>
      <a
        class="css-nk3r0j"
        href="/page/3"
      >
        »
      </a>
    </li>
  </ul>
</footer>
`;

exports[`Paginator with three pages on the last page renders correctly 1`] = `
<footer
  data-testid="paginator"
>
  <ul
    class="css-fm3157"
  >
    <li>
      <a
        aria-current="page"
        class="css-nk3r0j"
        href="/"
      >
        «
      </a>
    </li>
    <li>
      <a
        class="css-nk3r0j"
        href="/page/2"
      >
        ‹
      </a>
    </li>
    <li>
      <a
        aria-current="page"
        class="css-nk3r0j"
        href="/"
      >
        1
      </a>
    </li>
    <li>
      <a
        class="css-nk3r0j"
        href="/page/2"
      >
        2
      </a>
    </li>
    <li>
      3
    </li>
    <li>
      ›
    </li>
    <li>
      »
    </li>
  </ul>
</footer>
`;
